src/clr-angular/utils/drag-and-drop/draggable.ts

```diff
@@ -0,0 +1,246 @@
+        // So as both of these processes take place simultaneously by listening to the same event,
+        // the custom draggable inside this.dragStart() will be always undefined.
+
+        if (customDraggableGhost) {
+            this._customGhost = customDraggableGhost;
+            this.dragDispatcher.draggable.ghost = customDraggableGhost.draggableGhostEl;
+        }
+    }
+
+    ngAfterViewInit() {
+        this.dragDispatcher.draggable.self = this.el.nativeElement;
+        if (!this._customHandle) {
+            this.setDraggableStyles(this.dragDispatcher.draggable.self);
+        }
+        this.dragDispatcher.initialize();
+        this.subscriptions.push(this.dragDispatcher.onDragStart.subscribe(($event) => this.dragStart($event)));
+        this.subscriptions.push(this.dragDispatcher.onDragMove.subscribe(($event) => this.dragMove($event)));
+        this.subscriptions.push(this.dragDispatcher.onDragEnd.subscribe(($event) => this.dragEnd($event)));
+    }
+
+    ngOnDestroy() {
+        this.dragDispatcher.destroy();
+        this.subscriptions.forEach((sub: Subscription) => sub.unsubscribe());
+    }
+
+    private createDefaultGhost(originalNode: Node): Node {
+        const draggableGhost = this.viewContainerRef.createComponent(this.componentFactory, 0, this.injector,
+                                                                     [[originalNode.cloneNode(true)]]);
+        return draggableGhost.instance.draggableGhostEl;
+    }
+
+    private setDraggableStyles(originalNode: Node) {
+        this.renderer.addClass(originalNode, "draggable");
+        this.renderer.setAttribute(originalNode, "tabindex", "0");
+    }
+
+    private generateDraggableEvent(event: MouseEvent|TouchEvent): DraggableEvent {
+        let nativeEvent: any;
+        if ((<TouchEvent>event).hasOwnProperty("changedTouches")) {
+            nativeEvent = (<TouchEvent>event).changedTouches[0];
+        } else {
+            nativeEvent = event;
+        }
+
+        const clientPosition: DragPosition = {x: nativeEvent.pageX, y: nativeEvent.pageY};
+        return {
+            draggable: this.dragDispatcher.draggable,
+            dragPosition: clientPosition,
+            dragAndDropGroup: this.groupKey,
+            data: this.data
+        };
+    }
+
+    private dragStart(startEvent: MouseEvent|TouchEvent) {
+        const draggableEvent = this.generateDraggableEvent(startEvent);
+
+        this.dropSubscription = this.dragAndDrop.onDrop.subscribe(() => {
+            // Only Draggable which dragStart event detected on should subscribe to the drop event.
+            // That's why this subscription is set inside this method instead of AfterViewInit.
+            // And it has to be unsubscribed not only inside this.ngOnDestroy() as well as this.dragEnd() method.
+            this.drop();
+        });
+        this.subscriptions.push(this.dropSubscription);
+
+        if (!this.customGhostState.isDirectChild) {
```

Comment:
I've read every use of this `isDirectChild`, and I still have no idea at all why it's named like this. It seems to describe cases where the consumer provides a custom ghost, maybe?